src/pivot-table/data/extract-headers.ts

```diff
@@ -1,59 +1,111 @@
+import type { HeaderCell, HeadersDataMatrix, LayoutService, VisibleDimensionInfo } from "../../types/types";
 import getKey from "../components/helpers/get-key";
 
-const extractHeaders = (
+interface CreateMatrixProps {
+  layoutService: LayoutService;
+  lastCol: VisibleDimensionInfo[];
+  lastRow: VisibleDimensionInfo[];
+}
+
+const createEmptyMatrix = <T>(rowCount: number, colCount: number): (T | null)[][] =>
+  Array(rowCount)
+    .fill(null)
+    .map(() => Array.from({ length: colCount }, () => null));
+
+const transpose = <T>(matrix: T[][]) => matrix[0].map((_col, i) => matrix.map((row) => row[i]));
+
+const trimTrailingPseudo = (visibleDimensionInfos: VisibleDimensionInfo[]) => {
+  if (visibleDimensionInfos.at(-1) === PSEUDO_DIMENSION_INDEX) {
+    return visibleDimensionInfos.slice(0, -1);
+  }
+  return visibleDimensionInfos;
+};
+
+const createHeaderCell = (
   layoutService: LayoutService,
-  visibleTopDimensionInfo: VisibleDimensionInfo[],
-  visibleLeftDimensionInfo: VisibleDimensionInfo[],
-): (null | HeaderCell)[][] => {
+  qDimensionInfo: VisibleDimensionInfo,
+  colIdx: number,
+): HeaderCell => {
   const {
     layout: { qHyperCube },
     getDimensionInfoIndex,
   } = layoutService;
 
-  // rowCount cannot be 0, as it will cause an issue when there is no top data but there is left data
-  const rowCount = Math.max(1, visibleTopDimensionInfo.length);
-  const matrix: (null | HeaderCell)[][] = Array(rowCount)
-    .fill(null)
-    .map(() => Array.from({ length: visibleLeftDimensionInfo.length }, () => null));
-
-  visibleLeftDimensionInfo.forEach((qDimensionInfo, colIdx) => {
-    const id = getKey(qDimensionInfo);
-    if (qDimensionInfo === PSEUDO_DIMENSION_INDEX) {
-      matrix[rowCount - 1][colIdx] = {
-        id,
-        colIdx: -1,
-        label: "",
-        sortDirection: "A",
-        fieldId: "",
-        isActivelySorted: false,
-        isLocked: false,
-        isDim: false,
-        headTextAlign: "left",
-        dimensionInfoIndex: -1,
-      };
-    } else {
-      matrix[rowCount - 1][colIdx] = {
-        id,
-        colIdx,
-        label: qDimensionInfo.qFallbackTitle,
-        qReverseSort: qDimensionInfo.qReverseSort,
-        sortDirection:
-          qDimensionInfo.qSortIndicator && qDimensionInfo.qSortIndicator !== "N" ? qDimensionInfo.qSortIndicator : "A",
-        qLibraryId: qDimensionInfo.qLibraryId,
-        fieldId: qDimensionInfo.qGroupFieldDefs[qDimensionInfo.qGroupPos],
-        isActivelySorted: colIdx === (qHyperCube.activelySortedColumn?.colIdx ?? 0),
-        isLocked: qDimensionInfo.qLocked ?? false,
-        columnWidth: qDimensionInfo.columnWidth,
-        qApprMaxGlyphCount: qDimensionInfo.qApprMaxGlyphCount,
-        isDim: true,
-        headTextAlign: "left",
-        dimensionInfoIndex: getDimensionInfoIndex(qDimensionInfo),
-      };
-    }
+  const id = getKey(qDimensionInfo);
+  if (qDimensionInfo === PSEUDO_DIMENSION_INDEX)
```

Comment:
Lets add curly braces for the if statement

Reply:
Done